### ISSUE TITLE
Feature request: Nexus plugins

### ISSUE THREAD

**Issue by jferrettiboke** [prisma-labs/nexus#98]
I would LOVE to see support to create plugins in Nexus so that we can create plugins like [`nexus-prisma`](https://github.com/prisma/nexus-prisma).

I think this is the best way to do schema stitching within Nexus.

My current use case, as an example, is with Stripe. I am working on [`stripe-graphql`](https://github.com/jferrettiboke/stripe-graphql) and I would like to use it for some projects within Nexus to do schema stitching but in a more powerful way so that I can save time.

This "powerful way" allow us to:

- Select/hide only the fields we want.
- Rename fields.
- Override the resolver.

```ts
const Card = stripeObjectType({
  name: "Card",
  definition(t) {
    t.stripeFields([
      "id",
      "brand",
      "last4",
      { name: "exp_month", alias: "expMonth" },
      { name: "exp_year", alias: "expYear" }
    ])
  },
});
````

Pros that I see:

- It is easy to use/understand.
- It is more clear than the current ugly schema stitching.
- Nexus becomes more robust.
- Type-safe with TypeScript.
- With the ability to create plugins in Nexus, developers will be familiarized with the same syntax for any plugin.
- `nexus-prisma` won't have to maintain its own plugin, just the logic.

However, I got a question. How to use more than one plugin for the same object type? (I am not sure if there are important use cases for that.)

How many of you would be up for this? Let's see the support.

**Comment by tgriesser** [comment]
> I would LOVE to see support to create plugins in Nexus so that we can create plugins like nexus-prisma.

Actually considered this when building Nexus, (adding the ability to extend the `t.` namespace) and opted against it, in favor of composition. Prisma is sort of a special case in what they're doing with mapping a against a completely different generated GraphQL schema internally.

For cases like what you're suggesting you can usually just create "higher order functions" that expose a custom/configurable options object to achieve something like what you're going for:

```ts
const cardFieldTypeMapping = {
  id: 'ID',
  exp_month: 'Int',
  exp_year: 'Int',
  brand: 'String',
  last4: 'String'
}

interface TodoStripeTypes {
  // ... not taking the time to define these right now, 
  // but it should be fairly straightforward
} 

function stripeObjectType(cfg: TodoStripeTypes) {
   const { 
     definition, 
     cardFields = ['id', 'brand', 'last4', 'exp_month', 'exp_year'] 
   } = cfg
   return objectType({
      name: cfg.name,
      definition(t) {
        cardFields.forEach(field => {
          if (typeof field === 'string') {
            t.field(field, { type: cardFieldTypeMapping[field] })
          } else if (field && typeof field === 'object')  {
            Object.keys(field).forEach((key) => {
              t.field(field[key], { 
                type: cardFieldTypeMapping[key], 
                resolve: (root) => root[key] 
              })
            })
          }
        })
        if (typeof definition === 'function') {
          definition(t)
        }
      }
   })
}
```

then you can do:

```ts
const Card = stripeObjectType({
  name: "Card",
  cardFields: ["id", "brand", "last4", { exp_month: "expMonth" }, { exp_year: "expYear" }]
});
```

and optionally add new fields by providing a definition block:

```ts
const Card = stripeObjectType({
  name: "Card",
  cardFields: ["id", "brand", "last4", { exp_month: "expMonth" }, { exp_year: "expYear" }],
  definition(t) {
    t.field('filteredCharges', {
      type: 'SomeCustomType',
      list: true,
      resolve() {
        // ...
      }
    })
  }
});
```

> How many of you would be up for this? Let's see the support.

Before we use a popular vote to determine this feature, let's think more through the cases we're solving for and how they can be accomplished with the existing features of nexus. I do plan on simplifying some of the patterns used internally by `nexus-prisma` within nexus in order to help make the schema-composition work a little clearer than it does now.

**Comment by jferrettiboke** [comment]
@tgriesser Thanks for your detailed response. However, it is still confusing to me.

I would like that `stripe-graphql` exposes `stripeObjectType` so that I can use it in my projects. I already did this but I think I am missing something because I am not connecting the schemes. How can I do so? Does it require extra config? Right now I am getting an error which is "Error: prismaObjectType can only be used by `makePrismaSchema`". I have no idea about this.

PS: I know this issue is labeled as discussion but I would like to get it working so I can see pros and cons, and bring to this thread a more robust opinion.

**Comment by tgriesser** [comment]
Are you using `makePrismaSchema` or `makeSchema` to construct your schema?

**Comment by jferrettiboke** [comment]
Everything works well in the same repo (`stripe-graphql`) but this is not what I want. My issue is when I import `stripeObjectType` from `stripe-graphql` in another project (for example, `my-awesome-app`. How should I wire up both schemes? (I am using Yoga2 for the app.)

**Comment by Weakky** [comment]
@tgriesser I think there's no way easy way right now for jferrettiboke to accomplish what's he's trying to do.

If I'm understanding his problem well, he's trying to "merge" several schemas at the same time: One from prisma, and one from stripe.

However, afaik the current plugin architecture doesn't seem composable.

`nexus-prisma` has its own `makeSchema` function to get access to the prisma's graphql schema, but if you wanted to build your graphql server from yet another graphql source, then you'd have to build your own `PrismaStripeSchemaBuilder` and `makePrismaAndStripeSchema()` to read the schemas from both prisma and stripe.

TLDR: I think prisma's use-case is not special and we probably need a way to easily be able to compose these plugins and schemas.

On your example for instance, where would `stripeObjectType` read the schema from?

```
const Card = stripeObjectType({
  name: "Card",
  cardFields: ["id", "brand", "last4", { exp_month: "expMonth" }, { exp_year: "expYear" }],
  definition(t) {
    t.field('filteredCharges', {
      type: 'SomeCustomType',
      list: true,
      resolve() {
        // ...
      }
    })
  }
});
```

**Comment by jferrettiboke** [comment]
@Weakky I think you are very close to what I need. However, my use case is not about using Prisma (but other developers could have this use case where they need to mix both plugins at the same time).

I will provide more context so both of you can understand better my needs.

1. `stripe-graphql` is a Stripe GraphQL API which exports the schema, provide bindings and also a `stripeObjectType` function to be used within Nexus in order to simplify the process of connecting with Stripe (this is very similar to what `nexus-prisma` does).
2. Developers set up their own projects and use `stripe-graphql` as a third-party library.
3. If they use, `stripeObjectType`, they will avoid wasting time creating fields and resolvers. However, they can rename some specific fields and/or create some custom resolvers which could return data using the Stripe bindings provided by `stripe-graphql`.

The main use case for `stripeObjectType` is to simplify and save time to developers. So, instead of creating a whole entire structure of Stripe types and fields by their own, they can extend things very easy using `stripeObjectType`.

>On your example for instance, where would stripeObjectType read the schema from?

The ideal way would be that the `stripeObjectType` function could work without merging its own schema from `stripe-graphql` with the schema of the app, but I think this is not possible and they must be merged together at some point.

>And how would you access t.prismaFields() at the same time?

As I already said, I don't have that need yet. However, some developers could be interested in extending an object type from `stripeObjectType` but also from `prismaObjectType` at the same time. As a quick example, think in a `Customer` object type. They could want to expose fields from the database using `prismaObjectType` but also the cards field from `stripeObjectType`.

For more context, [here](https://github.com/jferrettiboke/stripe-graphql/tree/master/examples/my-app/src/graphql) is app example using Yoga2 with `stripeObjectType`. You have also [here](https://github.com/jferrettiboke/stripe-graphql/blob/master/src/nexus.ts#L43-L61) the `stripeObjectType` implementation for more context. Note that this is just a POC in order to illustrate better what my current use case is all about.

**Comment by jferrettiboke** [comment]
I have been thinking about how I would like to see this. Here's my proposal. What do you guys think?

```ts
 // Creates the `Customer` type in app schema
export const Customer = objectType({
  name: "Customer",
  definition(t) {
    // Refers to the `User` type in Prisma schema
    t.prismaObjectType({
      name: "User",
      fields: ["email"]
    });

    // Refers to the `Customer` type in Prisma schema
    t.prismaObjectType({
      name: "Customer",
      fields: ["stripeCustomerId"]
    });

    // Refers to the `Customer` type in Stripe schema
    t.stripeObjectType({
      name: "Customer",
      fields: [
        "id",
        "cards",
        {
          name: "created",
          alias: "createdAt"
        }
      ]
    });

    // Custom field
    t.field("subscription", {
      ...t.stripeType.subscription,
      resolve(root, args, ctx) {
        // Custom implementation
      }
    });
  }
});
```

Output:

```graphql
type Customer {
  email: String!
  stripeCustomerId: ID!
  id: ID!
  cards: [Card!]!
  createdAt: DateTime!
  subscription: Subscription
}
```

In this way, we can extend very easily without having to type by ourselves third-party schemes and/or fields.

**Comment by tgriesser** [comment]
Could this also be accomplished by having the `stripeObjectType` take the "builder" as an argument?

```ts
import { stripeType, stripeObjectType } from 'stripe-graphql'

 // Creates the `Customer` type in app schema
export const Customer = objectType({
  name: "Customer",
  definition(t) {
    // Refers to the `User` type in Prisma schema
    t.prismaObjectType({
      name: "User",
      fields: ["email"]
    });

    // Refers to the `Customer` type in Prisma schema
    t.prismaObjectType({
      name: "Customer",
      fields: ["stripeCustomerId"]
    });

    // Refers to the `Customer` type in Stripe schema
    stripeObjectType(t, {
      name: "Customer",
      fields: [
        "id",
        "cards",
        {
          name: "created",
          alias: "createdAt"
        }
      ]
    });

    // Custom field
    t.field("subscription", {
      ...stripeType.subscription,
      resolve(root, args, ctx) {
        // Custom implementation
      }
    });
  }
});
```

I believe this is the API I had originally suggested to Prisma specifically because it makes things a lot more difficult internally and from a type-generation standpoint to have things dynamically attached to the "builder" in the way that `nexus-prisma` does, and it's something I'll consider but I'm not sure it's something that will be generalized in the way you're describing in the short-term.

On a related note that might be relevant for your `stripe-graphql` project (which looks pretty neat btw), I personally find the merging of type fields from various schemas into a single type to be an anti-pattern, particularly when they're coming from external API sources like Stripe (with Prisma, you at least are in control of the data-source).

If I had a `Customer` object in my application, I would much prefer to have a field named `stripeCustomer` which would be an association to a `StripeCustomer` type rather than merging all of the fields into a single type, makes it clearer where the information is resolved from, simpler to cache, etc:

```ts
type Customer {
  email: String!
  stripeCustomer: StripeCustomer
}
```

**Comment by P4sca1** [comment]
I integrated nexus with ´graphql-shield` in the following way:
```ts
import {
  objectType,
  extendType,
  queryField,
  mutationField,
  subscriptionField,
} from 'nexus'
import {
  NexusObjectTypeConfig,
  NexusExtendTypeConfig,
  GetGen,
  FieldOutConfig,
  SubscribeFieldConfig,
} from 'nexus/dist/core'
import {
  ShieldRule,
  IRuleFieldMap,
  IRuleTypeMap,
} from 'graphql-shield/dist/types'
import { merge, capitalize } from 'lodash'
import { Rules } from './rules'
import { graphql as Log } from '../log'

export const isShieldRule = (
  rule: ShieldRule | IRuleFieldMap | IRuleTypeMap
): rule is ShieldRule => {
  // FieldMaps and TypeMaps have the Object constructor, whereas
  // Rule and LogicRule have a dedicated constructor (e.g. Rule, RuleAnd, RuleTrue, ...)
  return rule.constructor.name !== 'Object'
}

enum ParentType {
  NONE,
  QUERY,
  MUTATION,
  SUBSCRIPTION,
}

const mergeShieldRule = (
  type: ParentType,
  field: string,
  rule: ShieldRule | IRuleFieldMap
) => {
  let existingRule
  const fieldFullName =
    type === ParentType.NONE
      ? field
      : `${capitalize(ParentType[type])}.${field}`

  if (type === ParentType.NONE) {
    existingRule = Rules[field]
  } else {
    const accessor = capitalize(ParentType[type])

    // Query, Mutation and Subscription are already RuleFieldMaps.
    // Fields inside must be ShieldRules.
    // We can't have a RuleFieldMap inside a RuleFieldMap.
    if (!isShieldRule(rule)) {
      throw new Error(`Tried to assign RuleFieldMap to ${fieldFullName}`)
    }

    // Convert a ShieldRule on Query, Mutation or Subscription to a RuleFieldMap.
    // So { Query: allow } becomes { Query: { '*': allow }}.
    // This ensures the new field can correctly be added.
    const existingQMSRule = Rules[accessor]
    if (existingQMSRule === undefined) {
      Log.debug(
        `${accessor} type does not exist currently, initiating RuleFieldMap`
      )
      Rules[accessor] = {}
    } else if (isShieldRule(existingQMSRule)) {
      Log.debug(
        `${accessor} type is currently a ShieldRule, converting to a RuleFieldMap`
      )
      Rules[accessor] = { '*': existingQMSRule }
    }

    existingRule = (Rules[accessor] as IRuleFieldMap)[field]
  }

  // If there is no rule that currently exist, just assign the new rule.
  let newRule = rule

  if (!existingRule) {
    Log.debug(`Assigning ShieldRule to type ${fieldFullName}`)
  } else if (isShieldRule(existingRule)) {
    if (isShieldRule(rule)) {
      // If the existing rule and the new rule are both single rules that apply to the whole type, we will overwrite the existing rule with the new rule.
      Log.debug(
        `Type ${fieldFullName} already has an active ShieldRule. Overwriting with another ShieldRule`
      )
    } else {
      Log.debug(
        `Type ${fieldFullName} already has an active ShieldRule. Overwriting with a RuleFieldMap. The old rule will still be the default rule if no other default rule has been specified.`
      )

      newRule = merge({ '*': existingRule }, rule)
    }
  } else if (isShieldRule(rule)) {
    Log.debug(
      `Type ${fieldFullName} already has an active RuleFieldMap. Assigning new ShieldRule as default rule.`
    )
    newRule = merge(existingRule, {
      '*': rule,
    })
  } else {
    Log.debug(
      `Type ${fieldFullName} already has an active RuleFieldMap. Merging with new RuleFieldMap.`
    )
    newRule = merge(existingRule, rule)
  }

  if (type === ParentType.NONE) {
    Rules[field] = newRule
  } else {
    // We already ensured that Query, Mutation and Subscription are IRuleFieldMaps
    // and the new rule for the field is a ShieldRule.
    ;(Rules[capitalize(ParentType[type])] as IRuleFieldMap)[
      field
    ] = newRule as ShieldRule
  }
}

export const shieldedObjectType = <TypeName extends string>(
  config: NexusObjectTypeConfig<TypeName> & {
    rules?: ShieldRule | IRuleFieldMap
  }
) => {
  if (config.rules) {
    mergeShieldRule(ParentType.NONE, config.name, config.rules)
  }

  return objectType(config)
}

export const shieldedExtendType = <
  TypeName extends GetGen<'objectNames', string> | 'Query' | 'Mutation'
>(
  config: NexusExtendTypeConfig<TypeName> & {
    rules?: ShieldRule | IRuleFieldMap
  }
) => {
  if (config.rules) {
    mergeShieldRule(ParentType.NONE, config.type, config.rules)
  }

  return extendType(config)
}

export const shieldedQueryField = <FieldName extends string>(
  fieldName: FieldName,
  config:
    | FieldOutConfig<'Query', FieldName> & {
        rule?: ShieldRule
      }
    | (() => FieldOutConfig<'Query', FieldName> & {
        rule?: ShieldRule
      })
) => {
  if (typeof config === 'function') {
    // eslint-disable-next-line no-param-reassign
    config = config()
  }

  if (config.rule) {
    mergeShieldRule(ParentType.QUERY, fieldName, config.rule)
  }

  return queryField(fieldName, config)
}

export const shieldedMutationField = <FieldName extends string>(
  fieldName: FieldName,
  config:
    | FieldOutConfig<'Mutation', FieldName> & {
        rule?: ShieldRule
      }
    | (() => FieldOutConfig<'Mutation', FieldName> & {
        rule?: ShieldRule
      })
) => {
  if (typeof config === 'function') {
    // eslint-disable-next-line no-param-reassign
    config = config()
  }

  if (config.rule) {
    mergeShieldRule(ParentType.MUTATION, fieldName, config.rule)
  }

  return mutationField(fieldName, config)
}

export const shieldedSubscriptionField = <FieldName extends string>(
  fieldName: string,
  config:
    | SubscribeFieldConfig<'Subscription', FieldName> & { rule?: ShieldRule }
    | (() => SubscribeFieldConfig<'Subscription', FieldName> & {
        rule?: ShieldRule
      })
) => {
  if (typeof config === 'function') {
    // eslint-disable-next-line no-param-reassign
    config = config()
  }

  if (config.rule) {
    mergeShieldRule(ParentType.SUBSCRIPTION, fieldName, config.rule)
  }

  return subscriptionField(fieldName, config)
}
```

Sorry for the huge code block. I realised that I need to overwrite a bunch of nexus types (`objectType`, `queryField`, ...) to add the `rule` field to the config.
There are some problems with this approach though when it comes to combining it with other object type modifications.

It would be great if there was an api to register plugins, which can add fields to different object types and do something with it. The added fields could then be automatically added to the generated nexus typings for automatic type safety.

I haven't thought a lot about a good api to realise this. In my case it could look something like this:
```
import { registerPlugin } from 'nexus'

registerPlugin({
  objectType: {
    rule: {
      type: '' // Somehow get the type info here.
      onRegister: config => {
        // Do something with the  added field here.
      }
  }
})
```

**Comment by tgriesser** [comment]
Check out #143 - it's an initial pass at solving the problem of defining custom definition block fields.

**Comment by jferrettiboke** [comment]
@tgriesser I just checked out `dynamicOutputMethod` and play a bit with it. It's great to work within the same schema. However, I am not sure if I can delegate to another different schema.

To illustrate better my needs, I can give you an example of what I would like to achieve.

**stripe-graphql/src/StripeObjectType.ts**
```js
import { dynamicOutputMethod } from "nexus";
import generatedTypes from "../generated/mapped-types";

interface TypeFields {
  [key: string]: string | any;
}

interface Types {
  [key: string]: TypeFields;
}

const types: Types = generatedTypes;

interface Field {
  name: string;
  alias: string;
}

interface StripeObjectType {
  name: string;
  fields: Array<string | Field>;
}

export const StripeObjectType = dynamicOutputMethod({
  name: "stripeObjectType",
  typeDefinition: `(opts: {
    name: string,
    fields: Array<string | {
      name: string;
      alias: string;
    }>
  }): void`,
  factory({ typeDef: t, args: [{ name, fields }] }) {
    if (fields && fields.includes("*")) {
      fields = Object.keys(types[name]).map(key => key);
    }

    fields.forEach(field => {
      if (typeof field === "string") {
        t.field(field, { type: types[name][field] });
      } else if (field && typeof field === "object") {
        t.field(field.alias, {
          type: types[name][field.name],
          resolve: root => root[field.name]
        });
      }
    });
  }
});
```

**app/src/graphql/Query.ts**
```js
import { objectType, stringArg } from "nexus";
import { StripeObjectType as SOT } from "stripe-graphql";

export const StripeObjectType = SOT;

export const Query = objectType({
  name: "Query",
  definition(t) {
    // I DON'T WANT THIS! Time is gold.
    t.string("customer", {
      args: {
        id: stringArg()
      },
      async resolve(root, args, ctx) {
        return await stripe.customer({ id });
      }
    });

    // I WOULD LIKE SOMETHING LIKE THIS! Save me time and I will love you forever!

    // Expose all fields in `Query`.
    t.stripeObjectType({
      name: "Query",
      fields: ["*"]
    });

    // Expose only `customer` and `customers` fields in `Query`.
    t.stripeObjectType({
      name: "Query",
      fields: ["customer", "customers"]
    });

    // Rename `customer` field to `getCustomer` in `Query` and expose it.
    t.stripeObjectType({
      name: "Query",
      fields: [{ name: "customer", alias: "getCustomer" }]
    });

    // Expose all args for `customer` field in `Query`.
    t.stripeObjectType({
      name: "Query",
      fields: [{ name: "customer", args: ["*"] }]
    });

    // Rename `id` argument to `customerId` for `customer` field in `Query`.
    t.stripeObjectType({
      name: "Query",
      fields: [
        { name: "customer", args: [{ name: "id", alias: "customerId" }] }
      ]
    });

    // Custom field.
    t.field("customField", {
      ...t.stripeType.customers,
      resolve(root, args, ctx) {
        // Custom implementation
      }
    });
  }
});
```

Would something like this work? Can I merge parts of other schemes into my app schema? I just want to compose my app schema from third-party schemes easily and save time without having to mirror by myself everything.

**Comment by jferrettiboke** [comment]
@tgriesser I just found [this](https://github.com/Urigo/merge-graphql-schemas). This is what I need in Nexus. I would like to merge and extend my schema with other third-party schemes. Is this possible in Nexus today?

**Comment by jferrettiboke** [comment]
It seems like Hasura is trying to solve the same issue with “remote joins”. Having a similar feature in Nexus would be awesome. Please, let’s consider to prioritize this. We should have a easy way in Nexus to connect different with different data-sources. https://blog.hasura.io/remote-joins-a-graphql-api-to-join-database-and-other-data-sources/

**Comment by Weakky** [comment]
Hey 👋,

I'd like give a quick sum-up of the state of what we think plugins could be after some discussions with @tgriesser at GraphQL Conf:

In a sense, `dynamicOutputMethod` and `dynamicInputMethod` already provides a solid foundation for (type-safe) plugins. The next version of nexus-prisma already greatly benefits from that by simplifying a lot the surface API.

We think plugins can be used for 3 different use-cases:

1. The first one is **augmenting the `DefinitionBlock` capabilities** (the `t` param). (using `dynamicOutputMethod`, `dynamicInputMethod`, and eventually `dynamicOuputObjectProperty` (as explained in #161)).
These are useful to add any kind of additional behavior to your daily development workflow. `nexus-prisma` is now entirely built on that feature.

2. The second one is **augmenting type/field options**. A plugin could be used to add new options to type/field configurations. One immediate use-case that we thought about was middlewares. The whole middleware system could probably be developed as a nexus plugin

3. **CLI plugins**. This should probably be put in a separate issue, but nexus should soon have a CLI to help with daily workflows (such as scaffolding, generating types, `dev` command etc..), and we think the CLI itself should also be plugin-based to allow for third-party libraries to integrate with it. At Prisma, we think this plugin system could partly replace `yoga2`, by providing scaffolding capabilities specific to nexus-prisma

**Comment by blazestudios23** [comment]
Hello All, 

I'm also combining multiple GraphQL End Points, two of them being Prisma. 

But I have two endpoints that are external. I'm using GraphQL Binding to connect to them and bass their data in the context, however I can't figure out how to easily integrate them into Nexus. 

GraphQL-Binding has already created TS files, which I believe I just need to integrate into the makeSchema function:

`
import * as path from "path";
import Query from "./Query";
import { makeSchema } from "nexus";

const schema = makeSchema({
  types: [Query],
  outputs: {
    schema: path.join(
      __dirname,
      "../../generated/localord/server/localord.graphql"
    ),
    typegen: path.join(__dirname, "../../generated/localord/serverlocalord.ts")
  },
  typegenAutoConfig: {
    sources: [
      {
        source: path.join(__dirname, "../../generated/types.ts"),
        alias: "t"
      }
    ],
    contextType: "t.<No idea what to put here>"
  }
});

export { schema as default };
`

Is this even possible? I think that there has be be a way to use Nexus with Graphql-binding and seems like the easiest and best solution to this issue. 

Thanks for the help in advance.

**Comment by jasonkuhrt** [comment]
Related https://github.com/prisma/nexus/pull/200

**Comment by jasonkuhrt** [comment]
For anyone interested I have started to put together a proposal for plugins here https://www.notion.so/prismaio/Nexus-Plugin-System-cf2025b1f1084e4e98035c2acc2d3723.

**Comment by alidcastano** [comment]
is there support for generating dynamic root types (not just `t.field` but `objectType` outputs)? it seems to currently work with `dynamicOutputMethod` but not sure if the approach I took is correct or meant to be supported. 

my use-case is that I need to filter an objects nested relation fields by the same parent status. 

i.e. if I query a notebook` Notebook` that has `published: true` then for it's `Notes` I'd like to ensure only published ones are shown as well.

here's code for how I'm currently handling that:
```
// create a notebook factory using `DynamicOutputMethod`
dynamicOutputMethod({
  name: 'notebook',
  typeDefinition: `args: {
    typeName: string,
    fieldName: string,
    where: { published: boolean }
  }`,
 factory({ typeDef: t, args }) {
    args.forEach(({ typeName, fieldName, where: { published } }) => {
      const CustomNotebook = objectType({
        name: typeName,
        definition: t => {
          t.string('name')
          t.field('notes', { 
            type: 'Notes', 
            resolve: async (notebook, _, { db }) => db.Note.findMany({ 
              where: { notebookId: notebook.id, published },
            })
          })
        },
      })  

      t.field(fieldName, {
        type: CustomNotebook,
        args: { slugPath: stringArg() },
        resolve: (authUser, { id }, { db }) => db.Notebook.find({
          where: { id, ownerId: authUser.id, published }
        })
      })
    })
})

//  instantiate notebook object type as custom field definition 
objectType({
  type: 'Profile',
  definition: t => {
   t.notebook({
      typeName: 'PublishedNotebook',
      fieldName: 'publishedNotebook',
      where: { published: true }
    })
 }
})
```

**Comment by jasonkuhrt** [comment]
Hey @alidcastano thanks for sharing that. Your solution seems reasonable but I've noted your comment to think about later, and if there's a way the underlying API can better serve your use-case. It's interesting.

If you're motivated, please open a new feature request for an API that would better serve your use-case. Maybe we'll conclude that your approach is the idiomatic way but I think we need to talk that through more.

**Comment by tgriesser** [comment]
Closing, since [Plugins](https://nexus.js.org/docs/api-plugins) are now possible in the latest 0.12.0-rc.4. Feel free to open additional issues / PRs to improve the docs, they're a little bare bones at the moment.
